pkg/kv/kvserver/batcheval/cmd_end_transaction.go

```diff
@@ -1354,15 +1354,25 @@ func mergeTrigger(
 		if err != nil {
 			return result.Result{}, err
 		}
-		if lhsHint.Merge(rhsHint, rec.GetMVCCStats().HasNoUserData(), merge.RightMVCCStats.HasNoUserData()) {
-			updated, err := lhsLoader.SetGCHint(ctx, batch, ms, lhsHint)
-			if err != nil {
+
+		updated := false
+		leftEmpty, rightEmpty := rec.GetMVCCStats().HasNoUserData(), merge.RightMVCCStats.HasNoUserData()
+		// TODO(pavelkalinnikov): deprecate the cluster setting and call Merge()
+		// unconditionally when min supported version is 23.2.
+		if rec.ClusterSettings().Version.IsActive(ctx, clusterversion.V23_2) ||
+			enableStickyGCHint.Get(&rec.ClusterSettings().SV) {
```

Comment:
I think we should respect the new hint if it has already been written, regardless of the setting and gate. It's a lot easier to reason about, both for us and users, and can also be racy since version gates aren't enabled atomically across cluster nodes. Otherwise, existing hints will sometimes be respected but sometimes not, depending on whether there's been a range merge in the meanwhile.

Reply:
Sounds reasonable. Will modify the code as follows (if I understand the proposal right):
- Check the `cluster setting || version gate` condition only in `cmd_delete_range.go`, where the `GCHint` originates from.
- Here in `cmd_end_transaction.go`, only look at the new "GC timestamp" fields (to determine whether this is a "new" hint), and branch accordingly.

This has the "atomicity" benefit, which relies on the fact that if a GCHint has been written in the new format, then the node on which it "originated" had the gate enabled, which means that the entire cluster already uses the new binary which "understands" the new hint.

Reply:
> - Here in cmd_end_transaction.go, only look at the new "GC timestamp" fields (to determine whether this is a "new" hint), and branch accordingly.

Actually, sounds like we don't need to branch here. We just use the `Merge` function. It will automatically "respect" the new hints, but will also work for old hints.

Reply:
Yep, sounds good.

Reply:
Done.